snapcraft/storeapi/__init__.py

```diff
@@ -13,9 +13,285 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import hashlib
+import json
+import logging
+import os
 
-from .channels import get_channels, update_channels  # noqa
+from urllib import parse
+
+import requests
+from ssoclient import v2 as sso
+
+import snapcraft
+from snapcraft import config
+from .constants import (
+    DEFAULT_SERIES,
+    UBUNTU_SSO_API_ROOT_URL,
+    UBUNTU_STORE_API_ROOT_URL,
+    UBUNTU_STORE_SEARCH_ROOT_URL,
+)
+
+from . import _upload
 from .info import get_info  # noqa
-from ._login import login  # noqa
-from ._download import download  # noqa
-from ._upload import upload  # noqa
+
+
+logger = logging.getLogger(__name__)
+
+
+class InvalidCredentials(Exception):
+    pass
+
+
+class StoreError(Exception):
+    """Base class for all storeapi exceptions.
+
+    :cvar fmt: A format string that daughter classes override
+    """
+
+    fmt = 'Daughter classes should redefine this'
+
+    def __init__(self, **kwargs):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def __str__(self):
+        return self.fmt.format([], **self.__dict__)
+
+
+class SnapNotFound(StoreError):
+
+    fmt = 'The  "{name}" for {arch} was not found in {channel}.'
+
+    def __init__(self, name, channel, arch):
+        super().__init__(name=name, channel=channel, arch=arch)
+
+
+class SHAMismatch(StoreError):
+
+    fmt = 'SHA512 checksum for {path} is not {expected_sha}.'
+
+    def __init__(self, path, expected_sha):
+        super().__init__(path=path, expected_sha=expected_sha)
+
+
+def macaroon_auth(conf, acl):
```

Comment:
If the func is used only in this file, we start prepending the _. That helps figuring out which is the right entry point for a test, and gives us a lot of freedom when we are refactoring stuff. Once we need to access the func from outside, we make sure that it's clean enough for us to support it for the foreseeable future. So we delay that discussion about how ready is the code for long term consumption until there's an external consumer.